x11 forwarding

Joerger opened the pull request:
This PR implements [RFD](https://github.com/gravitational/teleport/pull/10009)

Closes https://github.com/gravitational/teleport/issues/7799

Joerger committed:
Add X11 forwarding flag to tsh ssh; Implement basic X11 forwarding client.

Joerger committed:
Implement x11 server.

Joerger committed:
Implement xauth with trusted and untrusted mode; move x11 code to sshutils/x11.

Joerger committed:
Update x11 forwarding flags and add openssh options.

Joerger committed:
Clean up implementation and add proper comments; reuse new x11 helpers
in forwarding server.

Joerger committed:
Add x11 server options; handle x11 request failures gracefully; Fix forward x11 timeout.

Joerger committed:
Fix x11 forwarding for XQuartz and MobaXTerm.

Joerger committed:
Update comment for ServerConfig.UseLocalhost.

Joerger committed:
Refactor xauth command logic to make the command run more customizable;
Set xauth entry within session re-exec rather than in the root process.

Joerger committed:
Update xserver unix listner owner in re-exec block.

Joerger committed:
Add x11rdy pipe to signal parent process when x11 is set up.

Joerger committed:
Update non-graceful restart TODO comment.

Joerger committed:
Clean up.

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+	}
+
+	log.Debug("creating an xauth cookie for x11 forwarding")
+	if xauthEntry == nil {
```
`xauthEntry` is always going to be nil at this point, right? Can we reduce some of the  nested if statements here?

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+		} else {
+			// Generate the xauth entry in a temporary file so
+			// it only exists within the context of this request.
+			xauthDir, err := os.MkdirTemp(os.TempDir(), "tsh-*")
```
```suggestion
			xauthDir, err := os.MkdirTemp("", "tsh-*")
``` 

The default behavior of `MkdirTemp` is to use `os.TempDir()` unless you specify a different dir by passing a non-empty string. Since you aren't overriding the default behavior I would make that explicit.

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+				return trace.Wrap(err)
+			}
+			defer os.RemoveAll(xauthDir)
+			xauthFile, err := os.CreateTemp(xauthDir, "xauthfile")
```
Do you even need to `MkdirTemp` at all? Why not just:

```
xauthFile, err := os.CreateTemp("", "xauthfile-***")
defer os.Remove(xauthFile.Name())
```

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+func (ns *NodeSession) serveX11Channels(ctx context.Context, sess *ssh.Session) error {
+	err := x11.ServeX11ChannelRequests(ctx, ns.nodeClient.Client, func(ctx context.Context, nch ssh.NewChannel) {
+		if !ns.x11RefuseTime.IsZero() && time.Now().After(ns.x11RefuseTime) {
+			nch.Reject(ssh.Prohibited, "rejected x11 channel request after ForwardX11Timeout")
```
Do you need to return here?

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+		var req x11.X11ChannelRequestPayload
+		if err := ssh.Unmarshal(nch.ExtraData(), &req); err != nil {
+			nch.Reject(ssh.Prohibited, "invalid payload")
+			log.Debug("rejected x11 channel request with invalid payload")
```
Include the error for debugging?

zmb3 commented on lib/client/session.go:
```diff
@@ -220,6 +246,169 @@ func (ns *NodeSession) createServerSession() (*ssh.Session, error) {
+		defer cancel()
+		go func() {
+			if err := sshutils.ForwardRequests(ctx, sin, sess); err != nil {
+				log.WithError(err).Debug("Failed to forward ssh request during x11 forwarding")
```
Should you check the error here and not log if it is `context.Canceled`?

Joerger replied:
if the context is canceled this should return nil for error

Joerger replied:
Adding this check inside `ForwardRequests` instead

zmb3 commented on lib/config/fileconf.go:
```diff
@@ -924,6 +928,34 @@ func (r *RestrictedSession) Parse() (*restricted.Config, error) {
+	DisplayOffset *uint `yaml:"display_offset,omitempty"`
+}
+
+// Parse will parse the enhanced session recording configuration.
```
1. Fix the godoc - this is not session recording configuration.
2. Please add test coverage showing valid/invalid configurations.

zmb3 commented on lib/events/api.go:
```diff
@@ -252,6 +252,11 @@ const (
 
+	// X11 forwarding channel open event
+	X11ForwardChannelEvent   = "x11-forward-channel"
+	X11ForwardChannelSuccess = "success"
```
Can you just use `X11ForwardSuccess` and `X11ForwardErr`? Why do we need two sets of constants with the same values?

zmb3 commented on lib/srv/authhandlers.go:
```diff
@@ -159,6 +159,14 @@ func (h *AuthHandlers) CheckAgentForward(ctx *ServerContext) error {
+// CheckX11Forward checks if X11 forwarding is permitted for the user's RoleSet.
+func (h *AuthHandlers) CheckX11Forward(ctx *ServerContext) error {
+	if !ctx.Identity.RoleSet.PermitX11Forwarding() {
+		return trace.Wrap(trace.AccessDenied("x11 forwarding not permitted"))
```
1. The other checkers return `trace.AccessDenied` directly, rather than wrapping it. Should you do the same?
2. Should you emit an audit event when access to X11 forwarding is denied?

zmb3 commented on lib/srv/ctx.go:
```diff
@@ -584,6 +603,128 @@ func (c *ServerContext) getSession() *session {
+		return trace.Wrap(err)
+	}
+
+	// TODO (Joerger): During a non-graceful shutdown, such as SIGTERM, the Teleport
```
This seems worth fixing sooner rather than later.

Joerger replied:
It's worth mentioning that this is also an issue with unix sockets opened for agent forwarding, so I'll work on a fix for non-graceful shutdown's separate from this PR.

zmb3 commented on lib/srv/forward/sshserver.go:
```diff
@@ -1086,64 +1006,45 @@ func (s *Server) handleX11Forward(ctx context.Context, ch ssh.Channel, req *ssh.
-		event.Status.Error = "x11 forwarding not permitted"
-		if err := s.EmitAuditEvent(s.closeContext, &event); err != nil {
+	defer func() {
+		if err := s.EmitAuditEvent(ctx, event); err != nil {
```
So this doesn't emit until the request has been forwarded?

Is there any potential for abuse here? Could someone activate X11 forwarding, do some things, and then get Teleport to crash before any event is emitted, covering their tracks?

Joerger replied:
`handleX11Forward` isn't called for every xserver request sent by the client (`xeyes`), it's only called on initial setup of the session, when the client sends the `x11-req` request during `tsh ssh - X`.

The actual x11 forwarding will be carried out by the goroutine spawned by `handleX11Forward`, and that goroutine won't start processing xserver requests until the session is fully set up. 

So in short, the event will always be emitted before x11 requests can be forwarded.

zmb3 commented on lib/srv/reexec.go:
```diff
@@ -153,12 +166,12 @@ func RunCommand() (io.Writer, int, error) {
 	var pty *os.File
 	uaccEnabled := false
 
-	// If a terminal was requested, file descriptor 4 and 5 always point to the
```
What's going on here?

The comment referred to fds 4 and 5, but the code referred to fds 5 and 6.

The comment was updated to refer to 5 and 6, but then the code was changed to 6 and 7.

??

Joerger replied:
Ah I fixed the comment first, then ended up incrementing it without adjusting the comment. Fixed.

zmb3 commented on lib/sshutils/x11/auth.go:
```diff
@@ -0,0 +1,221 @@
+	var protoLen, dataLen int
+	switch byteOrder := initBuf[0]; byteOrder {
+	///* Byte order MSB first. */
+	case 0x42:
```
make constants for these byte orders?

Joerger replied:
Changed these to 'B' and 'l' and simplified the switch statement.

zmb3 commented on lib/sshutils/x11/auth.go:
```diff
@@ -0,0 +1,221 @@
+		return nil, trace.Wrap(err)
+	}
+
+	// Replace auth data with the real auth data and write to conn
```
Can you use the `copy` builtin here?

zmb3 commented on lib/sshutils/x11/forward.go:
```diff
@@ -0,0 +1,229 @@
+// authProto and authCookie are required to set up authentication with the Server. screenNumber is used
+// by the server to determine which screen should be connected to for x11 forwarding. singleConnection is
+// an optional argument to request x11 forwarding for a single connection.
+func RequestX11Forwarding(sess *ssh.Session, display, authProto, authCookie string) error {
```
This will always be called with an `x11` prefix, so how about we name it `x11.RequestForwarding()`?

(And similar comment for `ServeX11ChannelRequests`)

zmb3 commented on tool/tsh/options.go:
```diff
@@ -163,14 +136,126 @@ type Options struct {
+		// support untrusted mode and will crash. For these reasons, users are encouraged
+		// to enable x11 forwarding with caution, as an attacker that can bypass file
+		// permissions on the remote host may be able to carry out such attacks.
+		ForwardX11Trusted: true,
```
Can we make sure this is covered in our docs as well? Seems too important to just be buried in the code.

zmb3 commented on tool/tsh/tsh.go:
```diff
@@ -383,6 +393,9 @@ func Run(args []string, opts ...cliOption) error {
 	ssh.Flag("option", "OpenSSH options in the format used in the configuration file").Short('o').AllowDuplicate().StringsVar(&cf.Options)
 	ssh.Flag("no-remote-exec", "Don't execute remote command, useful for port forwarding").Short('N').BoolVar(&cf.NoRemoteExec)
+	ssh.Flag("x", "Disable x11 forwarding for this request").Short('x').BoolVar(&cf.X11ForwardingDisabled)
+	ssh.Flag("X", "Enable x11 forwarding for this request").Short('X').BoolVar(&cf.X11Forwarding)
```
So there's a flag to enable and a flag to disable? What is the behavior when neither are set?

Joerger replied:
I'll remove `-x`, it makes sense in ssh where they have config files, but we only use flags and x11 forwarding is disabled by default.

zmb3 reviewed (commented):
Looking good but very few tests for the amount of code being added. Can we improve coverage a bit?

Joerger commented on lib/client/session.go:
```diff
@@ -75,6 +77,22 @@ type NodeSession struct {
+	x11RealCookie string
+	// x11FakeCookie is a spoof of the real cookie. This is used
+	// to authenticate x11 channel requests from the server without
+	// exposing the real cookie. The fake cookie is replaced with
```
There isn't an RFD because initially the protocol seemed straightforward enough to implement. After finding next to no documentation and delving into the OpenSSH source code, I found a lot more complexity. I am planning to add a lot more details to the PR description about authentication, security, etc. If you think it'd be better I can write an RFD about the implementation details instead.

Joerger committed:
Resolve PR comments and apply suggestions;

Add XConn and XDisplay abstractions;

Add tests in x11 package and sshserver;

Cleanup/fix forwarding logic;

Cleanup comments, fix X capitalization.

jakule commented on lib/srv/regular/sshserver.go:
```diff
@@ -1435,10 +1448,13 @@ func (s *Server) dispatch(ch ssh.Channel, req *ssh.Request, ctx *srv.ServerConte
 		return s.termHandlers.HandleWinChange(ch, req, ctx)
 	case sshutils.EnvRequest:
 		return s.handleEnv(ch, req, ctx)
+
```
nit: Do we need that extra line?

jakule commented on lib/sshutils/forward.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 Gravitational, Inc
```
nit:
```suggestion
// Copyright 2022 Gravitational, Inc
```

jakule commented on lib/sshutils/x11/auth.go:
```diff
@@ -0,0 +1,221 @@
+	if len(splitEntry) != 3 {
+		return nil, trace.Errorf("invalid xAuthEntry, expected entry to have three parts")
+	}
+	_, proto, cookie := splitEntry[0], splitEntry[1], splitEntry[2]
```
```suggestion
	proto, cookie := splitEntry[1], splitEntry[2]
```

jakule commented on lib/sshutils/x11/forward.go:
```diff
@@ -0,0 +1,229 @@
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		io.Copy(xconn, x11Chan)
```
The same as above. I don't think we should ignore errors here?

jakule commented on lib/sshutils/x11/forward.go:
```diff
@@ -0,0 +1,229 @@
+	}
+
+	splitDisplayNumber := strings.Split(splitHost[1], ".")
+	displayNumber, err := strconv.Atoi(splitDisplayNumber[0])
```
AFAIU this works with display numbers from 0-9. Any changes to go over that?

Joerger replied:
If `display=unix:10.1`, then `splitDisplayNumber=["10","1"], so this should work. I'll change the naming to make it less confusing

Joerger commented on lib/client/api.go:
```diff
@@ -2015,12 +2027,12 @@ func (tc *TeleportClient) runCommand(ctx context.Context, nodeClient *NodeClient
 
 // runShell starts an interactive SSH session/shell.
 // sessionID : when empty, creates a new shell. otherwise it tries to join the existing session.
```
Sorry for the confusion, I haven't committed the change yet, next commit will have everything covered.

Joerger committed:
Fix ForwardX11Timeout functionality;

Simplify X11 forwarding cli flags and options;

Make XServer proxy display max configurable;

Move and improve SSH options tests.

Joerger committed:
Add x11-untrusted-timeout flag.

Joerger committed:
Update UX to match updated RFD.

Joerger committed:
Fix re-exec chown syscall; Fix sshserver test.

Joerger committed:
Handle X Server listener limit exceeded error.

Joerger committed:
Remove todo.

Joerger commented:
@zmb3 @jakule PTAL, I addressed all of the comments except for the one about non-graceful shutdowns, as I'm going to open another PR for that bug fix since it's unrelated.

Joerger committed:
Cleanup.

zmb3 commented on lib/srv/ctx.go:
```diff
@@ -584,6 +603,136 @@ func (c *ServerContext) getSession() *session {
+}
+
+// getX11Config gets the x11 config for this server session.
+func (c *ServerContext) getX11Config() *X11Config {
```
Would it be better to return a copy `X11Config` instead of a pointer?

I'm wondering the caller can mutate the config without locking and introduce a race condition.

Joerger replied:
I've updated X11Config to contain no pointers and am returning a shallow copy here instead, is that fine?

zmb3 replied:
Should be.

zmb3 commented on lib/srv/reexec.go:
```diff
@@ -233,6 +246,62 @@ func RunCommand() (io.Writer, int, error) {
 
+	if c.X11Config != nil {
+		// Open x11rdy fd to signal parent process once X11 forwarding is set up.
+		x11rdyfd := os.NewFile(uintptr(5), "/proc/self/fd/5")
```
I see this is closed at the end of the block, but should you defer the close instead?

If an error occurs, this never gets closed - is that what you intend?

In fact, it looks like you don't write to `x11rdyfd` until the very end - can you move this code down closer to where it's needed so that you never even open it if an error is encountered?

zmb3 commented on lib/sshutils/x11/conn.go:
```diff
@@ -0,0 +1,79 @@
+	if displayOffset > maxDisplay {
+		return nil, Display{}, trace.BadParameter("displayOffset (%d) cannot be larger than maxDisplay (%d)", displayOffset, maxDisplay)
+	} else if maxDisplay > MaxDisplayNumber {
+		return nil, Display{}, trace.BadParameter("maxDisplay (%d) cannot be larger than the max int32 (2147483647)", maxDisplay)
```
You could use `math.MaxInt32` here rather than hard coding the number.

jakule replied:
One thing to point out. `maxDisplay` is an `int` which has 32 bits on 32 bits platforms. Technically we still support 32 bits. If someone pases a value that is larger than int32, `maxDisplay` will probably overflow and endup something < 0.

Joerger replied:
True, this edge case should lead to an error during file config unmarshaling or when this function is called, so it should be fine as is.

zmb3 commented on lib/sshutils/x11/display.go:
```diff
@@ -0,0 +1,196 @@
+func (d *Display) Dial() (XServerConn, error) {
+	tcpSock, tcpErr := d.tcpSocket()
+	if tcpErr == nil {
+		return net.DialTCP("tcp", nil, tcpSock)
```
What happens if `d.hostName` is `"unix"`?

It looks like it would be possible for it to try `net.DialTCP`, fail, and never try `DialUnix`. Am I getting that right?

Joerger replied:
I think `d.tcpSocket` would fail to resolve `unix:port`, but just in case of overlap I'll update Dial and Listen to try both and aggregate errors.

zmb3 commented on lib/sshutils/x11/display.go:
```diff
@@ -0,0 +1,196 @@
+	}
+
+	// Parse hostname.
+	colonIdx := strings.LastIndex(displayString, ":")
```
Nothing to change now, but I suspect `strings.Cut()` will be a nice cleanup here when Go 1.18 is out.

zmb3 reviewed (approved):
Looks good. A couple minor points, but nothing that should hold up a merge.

Joerger committed:
Fix linter and fileconf tests.

Joerger committed:
Check for EADDRINUSE when opening XServer unix socket.

Joerger committed:
Address comments.

Joerger committed:
Fix CI error.

jakule commented on lib/sshutils/x11/conn.go:
```diff
@@ -0,0 +1,90 @@
+	}
+
+	// Create /tmp/.X11-unix if it doesn't exist (such as in CI)
+	if err := os.Mkdir(x11SockDir(), 0777); err != nil && !errors.Is(err, os.ErrExist) {
```
On my Linux box `/tmp/.X11-unix` has permissions set to `1777`. Any reason why here we're using `0777`?

Joerger replied:
Ah I misread the file permissions, thanks for pointing this out.

jakule commented on tool/tsh/options.go:
```diff
@@ -163,46 +137,147 @@ type Options struct {
+	case "local":
+		o.ForwardAgent = client.ForwardAgentLocal
+	default:
+		return trace.BadParameter("invalid agent forwarding mode: %s", val)
```
```suggestion
		return trace.BadParameter("invalid agent forwarding mode %q, supported are: yes, no, local", val)
```

jakule commented:
Looks like one X11 test is currently falling in CI:
```
Step #1 - "run-tests": FAIL: github.com/gravitational/teleport/lib/srv/app
Step #1 - "run-tests": FAIL: github.com/gravitational/teleport/lib/srv/app.TestHandleConnection
Step #1 - "run-tests": FAIL: github.com/gravitational/teleport/lib/srv/regular
Step #1 - "run-tests": FAIL: github.com/gravitational/teleport/lib/srv/regular.TestX11Forward
```

Joerger committed:
Make suggested changes.

Joerger committed:
Continue session when forwarded x11 request denied; Print log to user when x11-req is denied.

Joerger committed:
Add debug log.

Fix broken test.

Check for xauth in tests.

Add xauth to buildbox.

Joerger committed:
Add xauth tests; Disable xauth tests unless requested.

Joerger committed:
Add todo for x11 test improvments.

Joerger committed:
Merge branch 'master' into joerger/x11-forwarding
